File: 0801-is-graph-bipartite/0801-is-graph-bipartite.cs
public class Solution {
    public bool IsBipartite(int[][] graph) {
        var n = graph.Length;
        var color = new int[n];
        Array.Fill(color, -1);

        for (int i = 0; i < n; i++) {
            if (color[i] == -1) {
                if (!(bfs(i, graph, color)))
                    return false;
            }
        }
        return true;
    }
    
    private bool bfs(int node, int[][] adj, int[] color) {
        var q = new Queue<int>();
        q.Enqueue(node);
        color[node] = 0;

        while (q.Count != 0) {
            int parent = q.Dequeue();
            foreach (var son in adj[parent]) {
                if (color[son] == -1) {
                    color[son] = 1 - color[parent];
                    q.Enqueue(son);
                }
                else if (color[son] == color[parent])
                    return false;
            } 
        }
        return true;
    }
}

File: 0801-is-graph-bipartite/0801-is-graph-bipartite.cpp
class Solution {
    using vi = vector<int>;

private:
    bool check(int node, vector<vi>& adj, vi& color) {
        queue<int> q;
        q.push(node);

        color[node] = 0;
        while (!q.empty()) {
            auto parent = q.front();
            q.pop();
            for (auto& son : adj[parent]) {
                if (color[son] == -1) {
                    color[son] = 1 - color[parent];
                    q.push(son);
                } else if (color[son] == color[parent])
                    return false;
            }
        }
        return true;
    }

public:
    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
        vi color(n, -1);

        for (int i = 0; i < n; i++) {
            if (color[i] == -1)
                if (!(check(i, graph, color)))
                    return false;
        }
        return true;
    }
};